vpr/src/tileable_rr_graph/tileable_rr_graph_gsb.cpp

```diff
@@ -1460,3 +1744,324 @@ void build_direct_connections_for_one_gsb(const RRGraphView& rr_graph,
+                    }
+                }
+                VTR_ASSERT(from.seg_index < seg_group.track_num * segment.length);
+                e_side side;
+                if (from_dir == 'W') side = RIGHT;
+                else if (from_dir == 'E') side = LEFT;
+                else if (from_dir == 'N') side = BOTTOM;
+                else if (from_dir == 'S') side = TOP;
+                else {
+                    VTR_LOGF_ERROR(__FILE__, __LINE__,
+                                   "Wrong segment from direction!\n");
+                    exit(1);
+                }
+
+                std::vector<size_t> track_list = rr_gsb.get_chan_node_ids_by_segment_ids(side, RRSegmentId(segment.seg_index));
+                if (track_list.size() == 0) continue;
+                else {
+                    VTR_ASSERT((int)track_list.size() >= (from.seg_index + 1) * 2);
+                    size_t seg_id;
+                    if (side == LEFT || side == BOTTOM) {  //INC
+                        seg_id = from.seg_index * 2;
+                    }
+                    else {                                 //DEC
+                        VTR_ASSERT(side == RIGHT || side == TOP);
+                        seg_id = from.seg_index * 2 + 1;
+                    }
+                    from_node = rr_gsb.get_chan_node(side, track_list[seg_id]);
+                    VTR_ASSERT(IN_PORT == rr_gsb.get_chan_node_direction(side, track_list[seg_id]));
+                    VTR_ASSERT(rr_gsb.is_chan_node(from_node));
+                }
+                    
+
+            }
+            else if (from.from_type == MUX) {
+                size_t from_mux_index = medium_mux_name2medium_index[layer][actual_coordinate.x()][actual_coordinate.y()][from.type_name];
+                from_node = rr_graph.node_lookup().find_node(layer, actual_coordinate.x(), actual_coordinate.y(), MEDIUM, from_mux_index);
+                VTR_ASSERT(rr_gsb.is_medium_node(from_node));
+            }
+            else {
+                VTR_LOGF_ERROR(__FILE__, __LINE__,
+                               "Wrong from type!\n");
+                exit(1);
+            }
+            VTR_ASSERT(from_node.is_valid());
+            auto iter = vib_map.begin();
+            for (; iter != vib_map.end(); ++iter) {
+                if (iter->first == from_node) {
+                    vib_map[from_node].push_back(to_node);
+                }
+            }
+            if (iter == vib_map.end()) {
+                std::vector<RRNodeId> to_nodes;
+                to_nodes.push_back(to_node);
+                vib_map.emplace(std::make_pair(from_node, to_nodes));
+            }
+                    
+            
+        }
+    }
+    /* Second stages*/
+    const std::vector<t_second_stage_mux_inf> second_stages = vib->get_second_stages();
+    for (size_t i_second_stage = 0; i_second_stage < second_stages.size(); i_second_stage++) {
+        std::vector<t_from_or_to_inf> froms = second_stages[i_second_stage].froms;
```

Comment:
Any difference between the first stage and second stage when checking the nodes and wires? If not, the big content inside the for-loop and be replaced by a common function. What do you think? Bug me if I was wrong.

Reply:
The checks for the first stage and the second stage are different, because their "from" and "to" node types are different.

Reply:
O.K. for now. We will refactor this part later.